third-party/realdds/include/realdds/dds-stream-server.h

```diff
@@ -1,61 +1,94 @@
-// License: Apache 2.0. See LICENSE file in root directory.
-// Copyright(c) 2022 Intel Corporation. All Rights Reserved.
-
-#pragma once
-
-#include <memory>
-#include <string>
-
-
-// Forward declaration
-namespace eprosima {
-namespace fastdds {
-namespace dds {
-class Publisher;
-class Topic;
-class TypeSupport;
-class DataWriter;
-}  // namespace dds
-}  // namespace fastdds
-}  // namespace eprosima
-
-
-namespace realdds {
-
-
-class dds_topic;
-class dds_topic_writer;
-
-
-struct image_header
-{
-    int format;
-    int height = 0;
-    int width = 0;
-
-    bool is_valid() const { return width != 0 && height != 0; }
-};
-
-
-// Distributes stream frames into a dedicated topic
-//
-class dds_stream_server
-{
-public:
-    dds_stream_server( std::shared_ptr< dds_topic_writer > const & topic );
-    ~dds_stream_server();
-
-    bool is_streaming() const { return _image_header.is_valid(); }
-    void start_streaming( const image_header & header );
-
-    void publish_image( const uint8_t * data, size_t size );
-
-    std::shared_ptr< dds_topic > const & get_topic() const;
-
-private:
-    std::shared_ptr< dds_topic_writer > _writer;
-    image_header _image_header;
-};
-
-
-}  // namespace realdds
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2022 Intel Corporation. All Rights Reserved.
+
+#pragma once
+
+#include "dds-stream-profile.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+
+namespace realdds {
+
+
+class dds_topic;
+class dds_topic_writer;
+class dds_publisher;
+class dds_stream_profile;
+
+
+struct image_header
+{
+    int format;
+    int height = 0;
+    int width = 0;
+
+    bool is_valid() const { return width != 0 && height != 0; }
+};
+
+
+typedef std::vector< std::shared_ptr< dds_stream_profile > > dds_stream_profiles;
+
+
+// Distributes stream images into a dedicated topic
+// 
+// This is a base class: you need to specify the type of stream via the instantiation of a video_stream_server, etc.
+//
+class dds_stream_server
+{
+    std::string const _name;
+    int const _default_profile_index;
+    dds_stream_profiles const _profiles;
+
+protected:
+    dds_stream_server( std::string const & name, dds_stream_profiles const & profiles, int default_profile_index );
+
+public:
+    virtual ~dds_stream_server();
+
+    std::string const & name() const { return _name; }
+    dds_stream_profiles const & profiles() const { return _profiles; }
+    int default_profile_index() const { return _default_profile_index; }
+
+    bool is_open() const { return !! _writer; }

```

Comment:
`!!`?

Reply:
"Multiple exclamation marks are a sure sign of a diseased mind", Terry Pratchett.    :-)

Reply:
In that case, the C++ standard is depraved.
This is what you do when there's no `operator bool()` but there is `operator!()`

Reply:
Not sure I understand.
You are trying to check if `_writer` is created (that is a call to `open()` was made), so you check the `shared_ptr`, but `shared_ptr` does have `operator bool()`

Reply:
Well it's `explicit` -- you can say `!! _writer` or `(bool) _writer` or even worse, `static_cast< bool >( _writer )`.

Reply:
or just `_writer.get() != nullptr`

But OK, I understand